### PR TITLE
NIFI-7103 Adding PutAzureDataLakeStorage Processor to provide native support for Azure Data Lake Storage Gen 2

### PR DESCRIPTION
**Pull request by MuazmaZ**
#### Description of PR

Added PutAzureDataLakeStorage Processor for Azure Data lake Storage Gen 2.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [X] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by turcsanyip** [comment]
@MuazmaZ Thank you for the contribution. This ADLS Gen 2 processor would definitely be a valuable addition to NiFi.
However, the CI builds failed. Could you please check them in the first round?

**Comment by MuazmaZ** [comment]
@turcsanyip added the missing dependency. All checks have passed.

**Comment by sburges** [comment]
Functionally this looks good, I built locally and tested against an Azure Data Lake account in Azure. The success and error cases work as expected.

@MuazmaZ One small thing, it looks like something happened to the indenting in the AbstractAzureDataLakeStorageProcessor class. It would be good to clean that up for readability.

**Comment by MuazmaZ** [comment]
Thanks @sburges for the feedback, indentation is fixed now

**Comment by sburges** [comment]
LGTM, thanks for addressing.

**Comment by MuazmaZ** [comment]
@pvillard31 thanks for the feedback.
- ADLS by default is encrypted with Microsoft Managed Keys. We can definitely add Key ID/Key Vault functionality for Customer Managed Keys for encryption as a feature add.
- Following Azure Blob Storage pattern we want to provide both options with Controller Service and with properties. 
I plan to add it along with permission (ACLs) as future additions. Let me know if it clarifies your questions.

**Comment by pvillard31** [comment]
Yes I was talking about CMEK. Totally fine with a follow-up JIRA.

Regarding the controller service, I don't see it as an option in this new processor. Here is the configuration for the Blob processor:

Screenshot 2020-03-11 at 23 36 51

Where Storage Credentials is the possible reference to the controller service. CS that I could configure with:

Screenshot 2020-03-11 at 23 37 10

With this new processor, I don't see the Controller Service option:

Screenshot 2020-03-11 at 23 36 41

Am I missing something?

**Comment by pvillard31** [comment]
Sorry I just read again your previous answer... you want to add it as a future addition. Please disregard my comments then ;)

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                        .buildClient();
+                } else if (StringUtils.isNotBlank(sasToken)) {
+                    storageClient = new DataLakeServiceClientBuilder().endpoint(endpoint).sasToken(sasToken)
+                        .buildClient();
+                } else {
+                    throw new IllegalArgumentException(String.format("Either '%s' or '%s' must be defined.",
+                        ACCOUNT_KEY.getDisplayName(), PROP_SAS_TOKEN.getDisplayName()));
+                }
+        final DataLakeFileSystemClient dataLakeFileSystemClient = storageClient.getFileSystemClient(fileSystem);
+        final DataLakeDirectoryClient directoryClient = dataLakeFileSystemClient.getDirectoryClient(directory);
```
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java]: Are there any exceptions that can be thrown out of this part of the code and should we be handling the `IllegalArgumentException`? I don't think whether `ACCOUNT_KEY` or `PROP_SAS_TOKEN` are defined will change between retries. It might be worth adding an IT test for that scenario to sort out whether it matters.

Comment (MuazmaZ) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java]: Thanks @jfrazee , added the IllegalArgumentException to capture the scenario.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml [hunk 2]
```diff
@@ -74,7 +74,7 @@
         <dependency>
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-core</artifactId>
-            <version>${jackson.version}</version>
+            <version>2.10.1</version>
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml]: Is this change required? If yes, can we use the latest (2.10.3)?
Ideally, I would try to remain consistent but I agree that changing the Jackson version in the root pom can be quite impacting...

Comment (MuazmaZ) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml]: Yes, it's required to use the latest sdk release. 2.10.1 matches the azure core sdk. I would stick to the azure version as compared to the latest. Let me know if it makes sense.

Comment (MuazmaZ) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml]: @pvillard31 could you let me know if this PR is good to be merged or you have additional comments?

Comment (pvillard31) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/pom.xml]: No, it's good on my side. If @turcsanyip is a +1, happy to merge.

**Comment by MuazmaZ** [comment]
@turcsanyip are you good with this? I would like to merge soon.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            fileClient.flush(length);
+
+            if (!attributes.isEmpty()) {
+                flowFile = session.putAllAttributes(flowFile, attributes);
+            }
+            session.transfer(flowFile, REL_SUCCESS);
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+            session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl().toString(), transferMillis);
+        } catch (IllegalArgumentException e) {
+            getLogger().error("Failed to create {}, due to {}", new Object[]{flowFile, e}, e);
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java]: "Failed to upload {FlowFile}" message may be more adequate. The FF already exists, so "Failed to create {FlowFile}" sounds a bit misleading for me.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            "Files that could not be written to Azure storage for some reason are transferred to this relationship")
+        .build();
+
+    private static final List <PropertyDescriptor> PROPERTIES = Collections.unmodifiableList(
+        Arrays.asList(AbstractAzureDataLakeStorageProcessor.ACCOUNT_KEY, AbstractAzureDataLakeStorageProcessor.ACCOUNT_NAME,
+            AbstractAzureDataLakeStorageProcessor.FILESYSTEM, AbstractAzureDataLakeStorageProcessor.DIRECTORY,
+            AbstractAzureDataLakeStorageProcessor.PROP_SAS_TOKEN, AbstractAzureDataLakeStorageProcessor.FILE));
+
+    private static final Set <Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet <>(
+        asList(
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java]: Could you please use either `Arrays.asList` or `asList` (with static import) both in lines 113 and 118?

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            attributes.put("azure.length", String.valueOf(length));
+            fileClient.flush(length);
+
+            if (!attributes.isEmpty()) {
+                flowFile = session.putAllAttributes(flowFile, attributes);
+            }
+            session.transfer(flowFile, REL_SUCCESS);
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+            session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl().toString(), transferMillis);
+        } catch (IllegalArgumentException e) {
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java]: All exceptions should be caught here. Eg. DataLakeStorageException coming from directoryClient.createFile(fileName) in line 93. Also exceptions from fileClient.append(in, 0, length) could be handled here.
So I would modify the class to Exception.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            " Instead of defining the Storage Account Name, Storage Account Key and SAS Token properties directly on the processor, " +
+            "the preferred way is to configure them through a controller service")
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .required(true)
+        .sensitive(true).build();
+
+    public static final PropertyDescriptor FILESYSTEM = new PropertyDescriptor.Builder().name("Filesystem-name")
+        .displayName("Filesystem Name")
+        .description("Name of the Azure Storage File System, It is assumed to be already existing")
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java]: Typo: `','` at the end of the first sentence. No `'.'` at the end of the second sentence.
The same for the Directory property.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .required(true)
+        .build();
+
+    public static final PropertyDescriptor DIRECTORY = new PropertyDescriptor.Builder().name("Directory-name")
+        .displayName("Directory Name")
+        .description("Name of the Azure Storage Directory, It is assumed to be already existing")
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java]: I think the non existing directories will be created. At least I did not get any errors.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                        ACCOUNT_KEY.getDisplayName(), PROP_SAS_TOKEN.getDisplayName()));
+            }
+            final DataLakeFileSystemClient dataLakeFileSystemClient = storageClient.getFileSystemClient(fileSystem);
+            final DataLakeDirectoryClient directoryClient = dataLakeFileSystemClient.getDirectoryClient(directory);
+
+            final long length = flowFile.getSize();
+            final Map<String, String> attributes = new HashMap<>();
+            final DataLakeFileClient fileClient = directoryClient.createFile(fileName);
+            if (length > 0) {
+                try (final InputStream rawIn = session.read(flowFile); final BufferedInputStream in = new BufferedInputStream(rawIn)) {
```
[the remark this answers is not in the document]
Comment (markap14) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java]: `InputStreamCallback` has existed from the beginning. The `InputStream ProcessSession.read(FlowFIle)` is a newer mechanism. Neither is necessarily preferred over the other. But the newer method, as used here, is generally easier to read, so I'd lean toward preferring that.

[the remark this answers is not in the document]
Comment (turcsanyip) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java]: @markap14 Thanks for the clarification. Then the current solution is appropriate, no need to change it.

**Comment by MuazmaZ** [comment]
@turcsanyip Thanks for your feedback. I have addressed all the comments. For high-level feedback:
- getRootDirectoryClient() function is not public in DataLakeFileSystemClient, I have reached out to the SDK team on this.
- Overwriting a file is also a follow-up with the SDK team as API supports a conditional If not match (https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create)
-  ACLs will be a follow-up task, already on the list. Some known issues need to be verified.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            attributes.put("azure.filename", fileName);
+            attributes.put("azure.primaryUri", fileClient.getFileUrl());
+            attributes.put("azure.length", String.valueOf(length));
+            flowFile = session.putAllAttributes(flowFile, attributes);
+
+
+            session.transfer(flowFile, REL_SUCCESS);
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+            session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl(), transferMillis);
+        } catch (IOException | StorageErrorException | IllegalArgumentException e) {
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java]: I'd recommend to catch `Exception` here.
I managed to throw `DataLakeStorageException` from the processor (eg. via specifying a non-existing filesystem or an invalid directory name on the UI). In this case the flowfile does not go to failure, but always goes back to the input queue of the processor which is not the expected behaviour I think (a wrong flowfile could hang up the processor, the other flowfiles would not be processed).
`DataLakeStorageException` is just an example I think so I would not add it as a fourth exception to be caught but rather catch `Exception`.
Or catch `(IOException | AzureException | IllegalArgumentException)`, but I'm afraid there can be other non-expected `RuntimeException-s` coming out from the library, so `Exception` seems to me the safest solution.

**Comment by MuazmaZ** [comment]
@turcsanyip confirmed with the SDK team that getDirectoryClient(“”) is valid. Made the change in property to allow empty string.
Removed PROP_ for SAS_TOKEN
Updated to catch Exception and tested the scenario. I was able to see the failure files with a non-existing filesystem.

**Comment by ravitejatvs** [comment]
@MuazmaZ I'm trying to push data to ADLS Gen2. But everytime, I get an exception that says `Failed to create file on Azure Data Lake Storage: com.azure.storage.file.datalake.models.DataLakeStorageException: Status code 400, "{"error":{"code":"InvalidUri","message":"The request URI is invalid.\nRequestId:3d7bfb40-b01f-0050-19d3-81c973000000\nTime:2020-09-03T09:17:20.1611855Z"}}"`. 

I am sure that the details provided are correct but still I am getting this. I also tried sending data to Storage Queue using the corresponding processor and that works fine. 

Please let me know if you need more details.
